packages/connect-query/src/message-key.ts

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021-2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { DescMessage, MessageInitShape } from "@bufbuild/protobuf";
+import { create } from "@bufbuild/protobuf";
+import type {
+  ReflectList,
+  ReflectMap,
+  ReflectMessage,
+} from "@bufbuild/protobuf/reflect";
+import { reflect } from "@bufbuild/protobuf/reflect";
+import { base64Encode } from "@bufbuild/protobuf/wire";
+
+/**
+ * For any given message, create an object that is suitable for a Query Key in
+ * TanStack Query:
+ *
+ * - Default values are omitted (both implicit and explicit field presence).
+ * - NaN, Infinity, and -Infinity are converted to a string.
+ * - Uint8Array is encoded to a string with Base64.
+ * - BigInt values are converted to a string.
+ * - Properties are sorted by Protobuf source order.
+ * - Map keys are sorted with Array.sort.
```

Comment:
Hmm, my assumption with tanstack keys is that they would have handled both of these without requiring explicit sorting being done, but I suppose best to be be sure.

Reply:
[hashKey does](https://github.com/TanStack/query/blob/9da00d610ab437ad1da6854c9a5697874ebebb7e/packages/query-core/src/utils.ts#L205), yes.

It's no extra cost for us to sort by Protobuf source order. I figure we might just as well sort map keys, and set a per-query `queryKeyHashFn: JSON.stringify`. No point in walking through the object yet another time.

Reply:
Yeah makes sense